scripts/ingest-tarball.sh

```diff
@@ -24,11 +31,151 @@ function echo_red() {
     echo -e "\e[31m$1\e[0m"
 }
 
+function echo_yellow() {
+    echo -e "\e[33m$1\e[0m"
+}
+
 function error() {
     echo_red "ERROR: $1" >&2
     exit 1
 }
 
+function check_version() {
+    if [ -z "${version}" ]
+    then
+        error "EESSI version cannot be derived from the filename."
+    fi
+
+    if [ -z "${tar_top_level_dir}" ]
+    then
+        error "no top level directory can be found in the tarball."
+    fi
+
+    # Check if the EESSI version number encoded in the filename
+    # is valid, i.e. matches the format YYYY.DD
+    if ! echo "${version}" | egrep -q '^20[0-9][0-9]\.(0[0-9]|1[0-2])$'
+    then
+        error "${version} is not a valid EESSI version."
+    fi
+
+    # Check if the version encoded in the filename matches the top-level dir inside the tarball
+    if [ "${version}" != "${tar_top_level_dir}" ]
+    then
+        error "the version in the filename (${version}) does not match the top-level directory in the tarball (${tar_top_level_dir})."
+    fi
+}
+
+function check_contents_type() {
+    # Check if the name of the second-level dir in the tarball matches to what is specified in the filename
+    if [ "${contents_type_dir}" != "${tar_contents_type_dir}" ]
+    then
+        error "the contents type in the filename (${contents_type_dir}) does not match the contents type in the tarball (${tar_contents_type_dir})."
+    fi
+
+    # Check if the second-level dir in the tarball is compat, software, or init
+    if [ ! -v content_types[${tar_contents_type_dir}] ]
+    then
+        error "the second directory level of the tarball contents should be either compat, software, or init."
+    fi
+}
+
+function cvmfs_regenerate_nested_catalogs() {
+    # Use the .cvmfsdirtab to generate nested catalogs for the ingested tarball
+    # ("cvmfs_server ingest" doesn't do this automatically)
+    echo "Generating the nested catalogs..."
+    cvmfs_server transaction "${repo}"
+    cvmfs_server publish -m "Generate catalogs after ingesting ${tar_file_basename}" "${repo}"
+    ec=$?
+    if [ $ec -eq 0 ]
+    then
+        echo_green "Nested catalogs for ${repo} have been created!"
+    else
+        echo_red "failure when creating nested catalogs for ${repo}."
+    fi
+}
+
+function cvmfs_ingest_tarball() {
+    # Do a regular "cvmfs_server ingest" for a given tarball,
+    # followed by regenerating the nested catalog
+    echo "Ingesting tarball ${tar_file} to ${repo}..."
+    ${decompress} "${tar_file}" | cvmfs_server ingest -t - -b "${basedir}" "${repo}"
```

Comment:
For `tar_file` it's already checked near the bottom of the script if it's set, and if it points to an existing file, so that can't be undefined here?

Reply:
I'm mainly pointing this out because at some point we could move this function elsewhere, and also use it on other contexts (where `$tar_file` may not be defined), and then bad things could happen.

It's mostly protecting ourselves against our future selves. ;)